### DISCUSSION TITLE
Inline relationship? Or how to customize renderer of relationship?

### DISCUSSION THREAD

**Discussion by Stupidism** [payloadcms/payload#586] [Feature Requests & Ideas]
I'm very interested with migrating to payloadcms now. Your solution for customization looks very promising. I'm still investigating/comparing the important features we're using with contentful. One of them is:

Our editors often inject some statistics data into our articles, whose value changes every day/week. Currently I'm using contentful inline entry to support this feature. I'd like to know how to do similar thing with payloadcms.

<img width="868" alt="image" src="https://user-images.githubusercontent.com/5801015/169997202-ab39e2d3-2ef8-431a-9dc8-a917ce711d9f.png">

> Of course, we can go back to text version like `{statistics.xxx}` just like when we use markdown beore.

**Comment by jmikrut** [comment]
Hey @Stupidism — you can actually do the same exact thing as an "inline entry" natively in Payload already!

It's the `relationship` element. Here is a screenshot from our [public demo](https://demo.payloadcms.com):

<img width="579" alt="image" src="https://user-images.githubusercontent.com/6651896/170168159-ce340349-fdd3-43fd-84c8-1fc7ec6f8e26.png">

[Here are the docs](https://payloadcms.com/docs/fields/rich-text#relationship-element) that talk about how to use this element.

**Reply by Stupidism** [reply]
I'm wondering if I can customize the render function of the `Post Relationship` Card in this rich text editor, so that I can change its css to `inline` to keep the paragraph whole.

**Comment by Stupidism** [comment]
@jmikrut I think I didn't make myself clear enough. I know relationships can be used as embedded blocks.

The key difference here is `inline elements`, like injecting a number, the current UI of relationship breaks the whole paragraph.

**Reply by jmikrut** [reply]
Ah I gotcha. You can build a custom inline element and have it allow for editors to build relationships, but the only drawback is that the relationship document(s) would have to be fetched manually or populated with a custom hook. Payload populates the `relationship` and `upload` element relations for you, but with a custom one, you'd need to do it yourself.

Look at the [public demo's video embed](https://github.com/payloadcms/public-demo/tree/master/src/fields/richText/video) for an example for how to build this, and then look at the SlateJS docs to [see how to make an element inline](https://www.slatejs.org/examples/inlines).

The above approach would be good because you'd be able to maintain both a block-based relationship, and you could have your own inline relationship component. 

Alternatively, if you wanted ALL of the relationship elements to be rendered inline, Payload would only need a small PR to achieve this. You'd need to be able to pass top-level plugins directly to the Slate editor, which is something that is not possible as of now. If we could pass top-level plugins, we could extend the Slate editor to consider `relationship` as inline.

```js
const withInlines = editor => {
  const { isInline } = editor

  editor.isInline = element => {
    return ['relationship'].includes(element.type) || isInline(element);
  }

  return editor
}
```

Then with some CSS overrides, we could render it as inline within the editor easily.

Honestly though if I were you I'd just go the first route, build a custom element and handle it that way. What do you think?

**Reply by Stupidism** [reply]
@jmikrut Thanks, I will try.
